### PR TITLE
Remove ICU library

### PR DESCRIPTION
**Pull request by dranikpg**
Replace ICU with unialgo

### CODE REVIEW

File: src/core/search/indices.cc [hunk 1]
```diff
@@ -8,12 +8,15 @@
...
 #include <absl/strings/numbers.h>
 #include <absl/strings/str_split.h>
-#include <unicode/brkiter.h>
-#include <unicode/unistr.h>
+
+#define UNI_ALGO_STATIC_DATA
+#define UNI_ALGO_DISABLE_NFKC_NFKD
+
+#include <uni_algo/case.h>
+#include <uni_algo/ranges_word.h>
```
Comment (dranikpg) [inline src/core/search/indices.cc:16]: We use add_third_party to include deps and use the headers directly ignoring its cmake so the defines need to be placed here.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -92,6 +92,12 @@ add_third_party(
   LIB "none"
 )
 
+add_third_party(
+  unialgo
+  URL https://github.com/uni-algo/uni-algo/archive/refs/tags/v1.0.0.tar.gz
+  LIB "none"
```
Comment (romange) [inline src/CMakeLists.txt:98]: do you prefer header only? this has potential to bloat an object code, no?
it is possible to build a static lib too.

File: src/server/search/search_family_test.cc [hunk 3]
```diff
@@ -361,6 +361,19 @@ TEST_F(SearchFamilyTest, Unicode) {
...
+  EXPECT_EQ(Run({"ft.create", "i1", "schema", "title", "text"}), "OK");
+
+  Run({"hset", "d:1", "title",
+       "WORD!!! Одно слово? Zwei Wörter. Comma before ,sentence, "
+       "Τρεις λέξεις: χελώνα-σκύλου-γάτας",
+       "visits", "400"});
+
+  // Make sure it includes ALL those words
+  EXPECT_THAT(Run({"ft.search", "i1", "word слово wörter sentence λέξεις γάτας"}),
+              AreDocIds("d:1"));
```
Comment (romange) [inline src/server/search/search_family_test.cc:374]: does it work with hebrew? :)

we do not have upper-case though.

`מה קורה קורה, אחי?`

Comment (dranikpg) [reply src/server/search/search_family_test.cc:374]: It does, I have a test above for unicode. This is just for word splitting, but added here one as well

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -92,6 +92,13 @@ add_third_party(
   LIB "none"
 )
 
+add_third_party(
+  uni-algo
+  URL https://github.com/uni-algo/uni-algo/archive/refs/tags/v1.0.0.tar.gz
+
+  BUILD_IN_SOURCE 1
+)
```
Comment (dranikpg) [inline src/CMakeLists.txt:100]: I looks so fragile but it freakin works

File: src/core/search/indices.cc [hunk 5]
```diff
@@ -8,12 +8,14 @@
 #include <absl/strings/ascii.h>
 #include <absl/strings/numbers.h>
 #include <absl/strings/str_split.h>
-#include <unicode/brkiter.h>
-#include <unicode/unistr.h>
+
+#define UNI_ALGO_DISABLE_NFKC_NFKD
```
Comment (romange) [inline src/core/search/indices.cc:12]: maybe add a comment explaining why you disabled this.

File: src/core/search/indices.cc [hunk 6]
```diff
@@ -8,12 +8,14 @@
...
+
+#include <uni_algo/case.h>
+#include <uni_algo/ranges_word.h>
 
 #include <algorithm>
 #include <cctype>
-#include <regex>
```
Comment (romange) [inline src/core/search/indices.cc:16]: halliluya

**Review by romange — APPROVED** [review]
Nice!

File: src/core/search/indices.cc (hunk 5 above)
Comment (dranikpg) [reply src/core/search/indices.cc:12]: Because its not needed, just so to reduce header parsing times 🤷🏻

**Pull request merged**